executor.go

```diff
@@ -186,27 +198,47 @@ func (e *Executor) verifySubjectAgainstReferrers(ctx context.Context, task *exec
 
 // verifyArtifact verifies the artifact by all configured verifiers and returns
 // error if any of the verifier fails.
-func (e *Executor) verifyArtifact(ctx context.Context, repo string, subjectDesc, artifact ocispec.Descriptor) ([]*VerificationResult, error) {
-	var verifierReports []*VerificationResult
-
+func (e *Executor) verifyArtifact(ctx context.Context, repo string, subjectDesc, artifact ocispec.Descriptor, artifactReport *ValidationReport, evaluator Evaluator) (*ValidationReport, error) {
+	artifactReport.Results = make([]*VerificationResult, 0)
 	for _, verifier := range e.Verifiers {
 		if !verifier.Verifiable(artifact) {
 			continue
 		}
-		verifierReport, err := verifier.Verify(ctx, &VerifyOptions{
-			Store:              e.Store,
-			Repository:         repo,
-			SubjectDescriptor:  subjectDesc,
-			ArtifactDescriptor: artifact,
-		})
-		if err != nil {
-			return nil, fmt.Errorf("failed to verify artifact %s@%s with verifier %s: %w", repo, subjectDesc.Digest, verifier.Name(), err)
-		}
 
-		verifierReports = append(verifierReports, verifierReport)
+		if evaluator != nil {
+			if err := evaluator.VerifyRequired(ctx, subjectDesc.Digest.String(), artifact.Digest.String(), verifier); err != ErrVerifyRequired {
+				continue
+			}
+			// Verify the subject artifact against the referrer artifact.
+			verifierReport, err := verifier.Verify(ctx, &VerifyOptions{
+				Store:              e.Store,
+				Repository:         repo,
+				SubjectDescriptor:  subjectDesc,
+				ArtifactDescriptor: artifact,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to verify artifact %s@%s with verifier %s: %w", repo, subjectDesc.Digest, verifier.Name(), err)
+			}
+			if evaluator.AddResult(ctx, subjectDesc.Digest.String(), artifact.Digest.String(), verifierReport) != nil {
+				return nil, fmt.Errorf("failed to add verifier report for artifact %s@%s verified by verifier %s: %w", repo, subjectDesc.Digest, verifier.Name(), err)
+			}
+			artifactReport.Results = append(artifactReport.Results, verifierReport)
+		} else {
+			// Verify the subject artifact against the referrer artifact.
+			verifierReport, err := verifier.Verify(ctx, &VerifyOptions{
+				Store:              e.Store,
+				Repository:         repo,
+				SubjectDescriptor:  subjectDesc,
+				ArtifactDescriptor: artifact,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to verify artifact %s@%s with verifier %s: %w", repo, subjectDesc.Digest, verifier.Name(), err)
+			}
+			artifactReport.Results = append(artifactReport.Results, verifierReport)
```

Comment:
This block of code is duplicated. Can you re-arrange the code or refactor a bit?